### PR TITLE
Fix ansible_virtualization_role fact on CloudLinux

### LINKED ISSUE #26424: ansible_virtualization_type on CloudLinux OS
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
facts

##### ANSIBLE VERSION
```
ansible 2.3.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['./modules', '/etc/ansible/modules/', '/usr/share/ansible']
  python version = 3.6.1 (default, Mar 24 2017, 12:50:34) [GCC 5.4.0]
```

##### SUMMARY
[This code](https://github.com/ansible/ansible/blob/e7deb07a87ea17881bef8d55890531a7f34f7d8e/lib/ansible/module_utils/facts/virtual/linux.py#L58-L64) tests for `/proc/vz` and sets `ansible_virtualization_type` to `openvz` when it exists.

The CloudLinux Distribution is actually not an OpenVZ system but both share some history which lead to the existance of `/proc/vz` on it. A [similar issue](https://projects.puppetlabs.com/issues/6728) was encountered and fixed for Puppet.

##### STEPS TO REPRODUCE
Install a [ISO](https://docs.cloudlinux.com/index.html?installing_new_servers.html) and gather facts on the installed host.

##### EXPECTED RESULTS
`ansible_virtualization_type` to be something different than `openvz`.

##### ACTUAL RESULTS
`ansible_virtualization_type` is `openvz`.

Here are some more facts from the machine I've tested on:
````json
"ansible_distribution": "CloudLinux",
"ansible_distribution_major_version": "6",
"ansible_distribution_release": "Igor Volk",
"ansible_distribution_version": "6.9",
"ansible_virtualization_role": "host",
"ansible_virtualization_type": "openvz",
````

### PR DESCRIPTION
**Pull request by silverwind**
##### SUMMARY
CloudLinux and OpenVZ have common roots, but CloudLinux does not really provide OS virtualization so it should not be regarded as a 'openvz' system. This change adds a check for the existance of the LVE kernel module which only exists on CloudLinux.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
facts

##### ANSIBLE VERSION
````
ansible 2.4.0 (cloudlinux-virtualization-type 297617fa80) last updated 2017/07/22 21:11:12 (GMT +200)
  config file = /Users/silverwind/git/ansible/ansible.cfg
  configured module search path = [u'/Users/silverwind/git/ansible-git/modules', u'/etc/ansible/modules', u'/usr/share/ansible']
  ansible python module location = /Users/silverwind/git/ansible-git/lib/ansible
  executable location = ./bin/ansible
  python version = 2.7.13 (default, Jul 18 2017, 09:17:00) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)]
````

##### ADDITIONAL INFORMATION
Fixes: https://github.com/ansible/ansible/issues/26424

### CODE REVIEW

**Comment by silverwind** [comment]
For comparision, here are the changed facts before and after this change:

### Physical machine:
##### Before:
```json
"ansible_virtualization_role": "host",
"ansible_virtualization_type": "openvz",
```

##### After:
```json
"ansible_virtualization_role": "NA",
"ansible_virtualization_type": "NA",
```

### Virtual Machine:
##### Before:
```json
"ansible_virtualization_role": "host",
"ansible_virtualization_type": "openvz",
```

##### After:
```json
"ansible_virtualization_role": "guest",
"ansible_virtualization_type": "VMware",
```

**Comment by silverwind** [comment]
I noticed that neither `/proc/vz` nor `/proc/lve` are visible for non-superusers, which will lead to wrong facts without `--become`.

Is it generally accepted that facts gathered as non-superuser might not be fully accurate? I'm not sure if the virtualization facts can even be gathered as a regular user.

**Comment by bcoca** [comment]
@silverwind yes, though it should be better documented, facts change depend on user that executes the gathering and of course root users get to "see more and more accurately" than normal ones.

**Comment by bcoca** [comment]
I would argue that openvz capable host should not be NA in your example above and that the original detection was working correctly.

**Comment by silverwind** [comment]
> I would argue that openvz capable host should not be NA

From what I gather, openvz hosting is possible, thought [limited](https://docs.cloudlinux.com/index.html?virtuozzo_and_openvz.html) and openvz cli tools are not preinstalled either. It's certainly not the main purpose of this distribution, which is per-user/process resource management for web hosting.

Also, puppet does a [pretty similar check](https://github.com/puppetlabs/facter/blob/b0726f54a5e8276ce756a126eb1fdd5b4fb08609/lib/src/facts/linux/virtualization_resolver.cc#L215-L218), so we'd be more compatible with them.

**Comment by silverwind** [comment]
@bcoca what do you think?

My motivation for this change is that I run playbooks that both target hosts running OpenVZ and Cloudlinux, and I have to add additional checks to not run OpenVZ specific tasks on CloudLinux because of this identification issue. I certainly would think that no one sane would event attempt to run OpenVZ containers on CloudLinux, given their beta (and most likely abandoned) status.

**Comment by alikins** [comment]
Given the puppet issue and it's similar change this seems reasonable to me.

Looking at other code that detects openvz, I notice virt-what also checks existence of /proc/bc
(http://git.annexia.org/?p=virt-what.git;a=blob;f=virt-what.in;h=9050035b526fefdc468f730f8c502e1e82156242;hb=HEAD#l151)
IIRC, it is distinguishing between running on a openvz host and a container running on a openvz host.
Do we want to check that as well? The current ansible virt info facts can't really express 'on a virt capable host but in a container so not really a virt host' so we should likely ignore /proc/bc for now.
so their may be no point in detecting it.

**Comment by alikins** [comment]
re /proc/bc, looks like systemd-detect-virt also checks for /proc/bc (https://github.com/systemd/systemd/blob/master/src/basic/virt.c#L418)

**Comment by silverwind** [comment]
The `/proc/bc` check [is already in Ansible](https://github.com/ansible/ansible/blob/0571014f2fe016f0055a61e56b1c56b129939420/lib/ansible/module_utils/facts/virtual/linux.py#L60-L63) and looks to be correctly used to distinguish host from guest. The only issue I see with it is that older OpenVZ kernels did not have this node, but `/proc/user_beancounters` instead. I guess this would be a topic for another issue.

So in summary, these are the `/proc` nodes present on the involved OpenVZ-based systems:

- OpenVZ Host: `/proc/vz`, `/proc/bc`
- OpenVZ Guest: `/proc/vz`
- CloudLinux: `/proc/vz`, `/proc/bc`, `/proc/lve`

As you can see, the `/proc/lve` check is neccessary to distinguish the OpenVZ host from CloudLinux.

**Comment by alikins** [comment]
@silverwind doh, I totally missed the already existing /proc/bc check in there. So the change looks good to me. LGTM

**Comment by alikins** [comment]
@silverwind 

Just to confirm, for the case:

> 
> 
> ### Virtual Machine:
> ##### Before:
> ```json
> "ansible_virtualization_role": "host",
> "ansible_virtualization_type": "openvz",
> ```
> 
> ##### After:
> ```json
> "ansible_virtualization_role": "guest",
> "ansible_virtualization_type": "VMware",
> ```

I'm assuming that VMware is the correct virt type ?  (this is for a cloudLinux vm running as VMware guest?)

**Comment by silverwind** [comment]
Yes, exactly. It's a CloudLinux guest running on a VMware host.
